### PR TITLE
JBPM-6001 - Trigger AdHoc Node is not implemented for stage, JBPM-600…

### PR DESCRIPTION
**Pull request by mswiderski**
…8 - Triggering ad-hoc fragments in specific case stage, added tests to cover close case from rules, from within process execution and from async command

### CODE REVIEW

**Comment by mswiderski** [comment]
Jenkins retest this

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/command/TriggerAdHocNodeInStageCommand.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public Void execute(Context context) {
+        KieSession ksession = ((RegistryContext) context).lookup( KieSession.class );
+        
+        ProcessInstance processInstance = ksession.getProcessInstance(processInstanceId);
+        
+        DynamicNodeInstance dynamicContext = (DynamicNodeInstance) ((WorkflowProcessInstanceImpl) processInstance).getNodeInstances(true).stream()
+                .filter(ni -> (ni instanceof DynamicNodeInstance) && stageId.equals(ni.getNode().getMetaData().get("UniqueId")))
+                .findFirst()
+                .orElse(null);
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/command/TriggerAdHocNodeInStageCommand.java]: Cannot this be changed to orElseThrow?

File: jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/impl/wih/AsyncWorkItemHandler.java [hunk 2]
```diff
@@ -29,6 +29,7 @@
 import org.kie.api.runtime.process.WorkItemHandler;
 import org.kie.api.runtime.process.WorkItemManager;
 import org.kie.api.runtime.query.QueryContext;
+import org.kie.internal.runtime.Cacheable;
```
Comment (MarianMacik) [inline jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/impl/wih/AsyncWorkItemHandler.java]: Just curious, does it provide any benefits to have WIH cacheable? Maybe performance improvement?

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseServiceImpl.java [hunk 3]
```diff
@@ -390,6 +392,37 @@ public void triggerAdHocFragment(Long processInstanceId, String fragmentName, Ob
         ProcessInstanceDesc pi = runtimeDataService.getProcessInstanceById(processInstanceId);
         internalTriggerAdHocFragment(pi, fragmentName, data);
     }
+    
+    @Override
+    public void triggerAdHocFragment(String caseId, String stageId, String fragmentName, Object data) throws CaseNotFoundException {
+        authorizationManager.checkAuthorization(caseId);
```
Comment (MarianMacik) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseServiceImpl.java]: This call is missing in the second method.

File: jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/impl/wih/AsyncWorkItemHandler.java (hunk 2 above)
Comment (mswiderski) [reply jbpm-services/jbpm-executor/src/main/java/org/jbpm/executor/impl/wih/AsyncWorkItemHandler.java]: it means that the work item handler does not have to be recreated for each execution - especially important for process instance and request strategies.

**Comment by mswiderski** [comment]
@MarianMacik all comments applied.

**Comment by mswiderski** [comment]
Jenkins retest this

**Pull request merged**